### ISSUE TITLE
[feature request] /etc/fstab.d

### ISSUE THREAD

**Issue by adrelanos** [karelzak/util-linux#790]
Quote "john.pseudonym1" <john.pseudonym1@protonmail.com>

> Hi,
> 
> As part of the hardening of an anonymity focused operating system called Whonix, we need to add different mount options for different filesystems e.g. hidepid=2 on /proc or noexec on /home. To make sure that a user's own fstab configurations are not messed up, we cannot use /etc/fstab for this. /etc/fstab.d would be the perfect thing for this but it seems that support for it has been [abandoned](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=663623). Will support for this ever be brought back? If not, are there any good alternatives?
> 
> Best regards.

[1] https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=928546

**Comment by karelzak** [comment]
The libmount allows to read fstab stuff from directory, for example
```
 mount --fstab /etc/fstab.d/
```
but this feature is not enabled by default and it does not check for fstab.d/ by default. (see "man mount" for more details).

The problem is that /etc/fstab is de-facto standard and on many places (in linux ecosystem) we do not expect this feature. The nice example is libc getmntent() API, another important is consumer systemd, /sbin/mount.<type> helpers, many 3rd party scripts etc ... so it's not about util-linux.

This is reason why libmount does not support it natively (although I agree that /etc/fstab.d/ itself is a good idea).

**Comment by adrelanos** [comment]
Karel Zak:
> The problem is that /etc/fstab is de-facto standard and on many places (in linux ecosystem) we do not expect this feature. The nice example is libc getmntent() API, another important is consumer systemd, /sbin/mount.<type> helpers, many 3rd party scripts etc ... so it's not about util-linux.

To find a way out of deadlock for backwards compatibility...

What do you think about adding support for this but having it disabled
by default? Third parties such as systemd could then add optional
support for it. Once widely supported, linux distribution decide if they
like to make the switch and enable this feature by default.

Or am I understanding this wrong, and there already is /etc/fstab.d in
util-linux and I'd have to suggest /etc/fstab.d support to libc,
systemd? Who'd be the recipient for /sbin/mount.<type> helpers?

**Comment by adrelanos** [comment]
[Was suggested to systemd.](https://github.com/systemd/systemd/issues/12506) This is the answer we got:

Quote @poettering https://github.com/systemd/systemd/issues/12506#issuecomment-490227369

> We don't define the /etc/fstab format, util-linux does that and we noawadays use util-linux' apis to parse it. Hence if you want soething like that, you'd have to work with the util-linux folks, and then ths would magically just work in systemd.
> 
> That said, if I were you I'd just write a generator that looks at /etc/fstab and creates drop-ins for the .mount units systemd-fstab-generator generates that override the Option= line. i.e. let systemd-fstab-generator do its thing as it currently does (meaning: translating fstab into native systemd .mount unit files), but then have your own generator that tweak some of these mount units with additional generated drop-ins as you need.
> 
> Anyway, let's close this, as I don't think we should change this in systemd upstream. I hope that makes sense.

**Comment by karelzak** [comment]
Yes, we can add --enable-fstabdir easily, the question is how usable it will be... (see my comment in systemd/systemd#12506). I'll try to prepare a patch.

We need to move this discussion to util-linux mailing list.

**Comment by karelzak** [comment]
I've created topic/fstab.d branch, we can later merge it to master (for example for v2.35).

**Comment by marcosfrm** [comment]
Any news? systemd/systemd#12506 was closed, but it still needs porting from setmntent() to libmount. Once this is in upstream systemd it will be easier to test util-linux changes.

**Comment by temporaryrespite** [comment]
There's commit 9354fea34283f54e45fb7d6faa7b54f516081780
Date:   Tue Jan 24 11:16:53 **2012** +0100
Author: Karel Zak
```
    libmount: remove fstab.d support
    
    If you really need to store your configuration to a alternative files
    then you have to explicitly specify the files or directories on
    mount(8) command line by --fstab option.
    
```

yes that's from 2012 :) and this is the only thing I've found about fstab.d on ML: https://marc.info/?l=util-linux-ng&m=132740311801201&w=2  and it's from 2012 also.

So, I wanted to ask something: if this fstab.d gets reimplemented, then how would multiple entries be handled? for example, there could be two mount entries for `/tmp` one as `zram` and one as `tmpfs`, both in /etc/fstab.d/ in two different files, let's say. 

What I actually want right now, is a way to have them both exist in that order, and have `mount` move to the next one and try it  if the former one(s) failed for whatever reason (eg. no zram support, or device, at the time the `mount` was executed).

Is there already an issue/feature requesting this, even if not for /etc/fstab.d/ but for /etc/fstab instead ? In other words: have multiple entries and have them each tried in order until one succeeds. I guess it might make things harder for `--options-mode` too, maybe (haven't thought about it much)

**Comment by karelzak** [comment]
On Thu, Apr 02, 2020 at 06:31:58PM -0700, gentoo_eshoes wrote:
> There's commit 9354fea34283f54e45fb7d6faa7b54f516081780
> Date:   Tue Jan 24 11:16:53 **2012** +0100
> Author: Karel Zak
> ```
>     libmount: remove fstab.d support
>     
>     If you really need to store your configuration to a alternative files
>     then you have to explicitly specify the files or directories on
>     mount(8) command line by --fstab option.
>     
> ```
> 
> yes that's from 2012 :) and this is the only thing I've found about
> fstab.d on ML:
> https://marc.info/?l=util-linux-ng&m=132740311801201&w=2  and it's
> from 2012 also.
> 
> So, I wanted to ask something: if this fstab.d gets reimplemented,

I need to talk about it with systemd guys. Not sure what is the
current status with setmntent().

Frankly, there was always many people who don't like fstab.d idea
because it makes a change to very old standard...

> then how would multiple entries be handled? 
> for example, there could
> be two mount entries for `/tmp` one as `zram` and one as `tmpfs`,
> both in /etc/fstab.d/ in two different files, let's say. 

The winner is the first entry.

IMHO your feature request is not related to fstab.d.

> What I actually want right now, is a way to have them both exist in
> that order, and have `mount` move to the next one and try it  if the
> former one(s) failed for whatever reason (eg. no zram support, or
> device, at the time the `mount` was executed).
> 
> Is there already an issue/feature requesting this, even if not for
> /etc/fstab.d/ but for /etc/fstab instead ? In other words: have
> multiple entries and have them each tried in order until one
> succeeds.

It would be possible to support this feature by a new option,
something like

    mount --next-on-fail /mnt

for fstab

    /dev/sda1   /mnt    auto    rw,noauto
    /dev/sda2   /mnt    auto    rw,noauto
    /dev/sda3   /mnt    auto    rw,noauto

or introduce a new X-mount.next_on_fail mount options.

But do we really need it? ;-) What is your use-case?

I can imagine that for example for NFS in some large infrastructure it 
sounds like a good fail-over solution if the same stuff is provided by
multiple server, or so.

    Karel

-- 
 Karel Zak  <kzak@redhat.com>
 http://karelzak.blogspot.com

**Comment by temporaryrespite** [comment]
> IMHO your feature request is not related to fstab.d.

you're right, let's not discuss it anymore(in this thread at least), besides, I've already edited [the comment](https://github.com/karelzak/util-linux/issues/790#issuecomment-608179868) with the patch that I'm using for my use case(just wanted to mount /tmp as zram and if fail then tmpfs)[you couldn't see the edits because you used email to reply]
